at_cmds-1/setdbg.tproj/setdbg.8
.TH setdbg 8
.SH NAME
.B setdbg 
- AppleTalk interactive kernel debugging facility. 

.SH SYNOPSIS
.B setdbg 
[
.I debugfile
]

.SH ARGUMENTS
.TP 8
.I debugfile 
Specifies the name of the AppleTalk kernel debugging
definition (usually netat/h/debug.h) file.  
If you don't specify the debugging file, setdbg will
look at the AT_DEBUG_DIR variable to find the "debug.h"
kernel debugging definitions. If this fails, setbdg
will look for "debug.h" in the current directory. Note
that the definition file used has to be the same as
the one user to compile the kernel running on the 
system where the traces are needed.

   setdbg myATdebugdefinitions.h


.SH DESCRIPTION

.B setdgb 
reads the debugging #defines as used in the kernel 
inside AppleTalk source code to select the level of traces
generated by different AppleTalk components on the debugging port
(using kprintf). Once started, setdbg will display a list of
all the trace level (usually starting with "L_") and all the
debugging modules (starting with "M_") that can be turned on
or off. A debug component selected for tracing will show a
"*" following its number. Note that the kernel debugging option
needs to be turned on for kprintf in order to generate any AppleTalk
traces. 

The following commands can be used within setdbg:  

   [number] - toggle the tracing of one component
   q - quit setdbg without updating the running 
       kernel with the new debugging settings.
   x - eXit and update the running kernel with the
       new debugging settings 
   t - toggle state of all debugging levels and modules 
   u - update the kernel immediately
   l - list show the kernel debugging status
   s - set all components to trace ON
   c - clear all components to trace OFF 

.SH FILES
.TP 8
/usr/local/bin/setdbg
Executable file
.TP 8
debug.h 
AppleTalk debugging header definitions


objc-1/ChangeLog
Changes from Ali to all the string class files. (12/23/92)
Changes from Derek to deal with name conflicts.

Cleaned up some warnings (7/23/91):

	* Makefile (COMMONFLAGS):  Changed -Wimplicit to -Wall.
	Let's make this compile	without warnings!

	* List.m, Storage.m (read:, write:):  Changed to return self.

	* <various files>:  Removed all unused local variables.

Added support for string objects (7/23/91):

	* NXString.m, NXCharacterSet.m, unichar.h, StringTest.m:  New files.

	* Makefile (LOCAL_CLASSES):  Added NXString.m and NXCharacterSet.m

	* Makefile (LOCAL_INCFILES):  Added unichar.h

	* objc-runtime.m (_objc_fixup_string_object_class):  New routine.
	If NXConstantString is present in class hash table, copy it to
	_NXConstantString_class and insert this address in class hashtable.

	* objc-runtime.m (_objcInit):  Call new routine after classes have
	been fixed up (in _objc_map_selectors).

	* objc-globaldata.m (_NXConstantString_class):  New global variable.
	Reduce pad by 36 bytes to compensate.

	* spec_sys:  Added slots for NXCompareCharacters(), NXEmptyString(),
	and NXFindCharacters().  Added files NXString.o and NXCharacterSet.o
	These changes will need to be integrated into libsys's copy.

Changed -[Object copy] (7/24/91):

	* Object.m (copy):  Chnaged to return [self copyFromZone: [self zone]]
	rather than using _copy.  The function object_copy() is no longer used
	by the runtime system, and could probably be removed.

	* List.[hm] (copy):  Deleted redundant method.

	* Storage.[hm] (copy):  Deleted redundant method.

	* HashTable.[hm] (copy):  Deleted redundant method.

	* NXString.[hm] (copy):  Deleted redundant method.

	* NXCharacterSet.[hm] (copy):  Deleted redundant method.

Changed method names in Storage (7/24/91):

	* Storage.[hm] (replaceElementAt:with:)  Changed from replace:at:
	and added 2.0 compatibility method in implementation only.

	* Storage.[hm] (insertElement:at:)  Changed from insert:at:
	and added 2.0 compatibility method in implementation only.

	* Storage.[hm] (removeElementAt:)  Changed from removeAt:
	and added 2.0 compatibility method in implementation only.

Cleaned up Makefile slightly (7/24/91):

	* Makefile (BY_PRODUCTS):  Added objcopt so that make clean removes
	all files not installed by make installsrc.

Fixed bug #14923 in HashTable (7/24/91):

	* HashTable.m (nextState:key:value:):  Set key and value to NULL
	when returning NO.

Fixed bug #14309 in Storage (7/24/91):

	* Storage.m (setNumSlots:):  Changed incorrect test
	(numSlots > maxElements) to (maxElements > old).

	* Storage.m (setNumSlots:):  Changed to use a single realloc.

Fixed bug #5687 in hashtable (7/24/91):

	* hashtable.h (NXHashInsertIfAbsent):  Improved comment.

Added appendList: method (dmoore 7/24/91)

	* List.m, List.h added method.
	
Added maptable functionality (dmoore 7/24/91)

	* maptable.c maptable.m

Removed typedef for id when compiling Objective-C (7/26/91):

	* objc.h (id):  Do not define typedef for id when compiling
	Objective-C (when __OBJC__ is predefined).
	
Made Objective C check for duplicate classes and use the one in the application
	(dmoore 7/27/91)
	
	* objc-runtime.m:  Added following functions:
		_objc_inform:	Just prints message for now.
		_resolveClassConflict: Makes sure that old class goes back in
			if it's from the app.
		_nameForHeader:	Returns the name of an image.
		_getExecHeader:	Picks out the app header.
		_headerForClass: Finds the mach_header containing a class.
		
		Modified _objc_get_classes to check for duplicates.
		
Added maptable prototypes to objc-globaltext.m (dmoore 7/27/91)

Modified spec_sys to include NXString and maptable stuff.

snaroff (7/31/91):

Integrate blaines kernel mods in a rational way.

Renamed "_objc_internal_error" to "_objc_inform"...removed "_objc_inform",
they were nearly identical (snaroff 7/31/91)

Create file "objc-errors.m"...it includes _objc_fatal, _objc_inform,
__objc_error, and _objc_error. MUST ADD _objc_inform and NXLogError to list 
of private externs!!!!!

Call _objc_fatal() instead of perror() in objc-sel.m...callings perror in these situations is bogus.

Have hashtable use NXLogError instead of fprintf(). 

Add field to objc_class for protocols...add protocol templates to objc-class.h

Installed new NXString files from ali (8/3/91):

	* NXString.[hm]:  Most methods are now statically typed.
	
	* NXStringTest.m  Changes to match.

Changes for clean compilation with new compiler (8/3/91):

	* Makefile (objcopt):  Use -ObjC flag with objc-edit.c

	* objcedit.c (round):  Made declaration global as well as definition.

	* objcedit.c (map_input):  Cast third arg to map_fd().

	* objcopt.c (get_objc):  Cast to (SEL *) two places.

	* objc-sel.m (_sel_writeHashTable):  Cast to (POHASH) two places.

	* objc-class.m (cache_flush, _cache_expand, _cache_flush):
	Made iteration variable unsigned.

	* hashtable.m (CopyIntoReadOnly):  Made iteration variable unsigned.

	* typedstream.m (InternalWriteObject):  Cleaned up printf format
	specifiers to match args.

	* typedstreamprivate.h:  Changed all function declarations to static.

	* objc-utils.m:  Cleaned up printf format specifiers to match args.

	* NXString.m (getBytes:range:, getCharacters:range:, NXFindCharacters,
	find:range:flags:usingTable:, _debugPrint):
	Made iteration variable unsigned.

	* NXCharacterSet.m (init, addCharacters:length:,
	removeCharacters:length:):  Made iteration variable unsigned.

Improvements to NXConstantString's (8/12/91):

	* NXString.[hm] (new):  Add version which raises exception.

	* NXString.[hm] (copyFromZone:):  Return an NXReadOnlyString copy
	instead of self, since our code might be unloaded.

	* NXString.[hm] (write:):  Write an NXReadOnlyString copy.

	* NXString.[hm] (finishUnarchiving):  Removed.

	* NXString.[hm] (free):  This now raises an error, since copyFromZone:
	no longer returns self.  You never allocate NXConstantString's, so you
	should never free them.


at_cmds-1/setdbg.tproj/setdbg.c
/*
 * Copyright (c) 1999 Apple Computer, Inc. All rights reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * "Portions Copyright (c) 1999 Apple Computer, Inc.  All Rights
 * Reserved.  This file contains Original Code and/or Modifications of
 * Original Code as defined in and that are subject to the Apple Public
 * Source License Version 1.0 (the 'License').  You may not use this file
 * except in compliance with the License.  Please obtain a copy of the
 * License at http://www.apple.com/publicsource and read it before using
 * this file.
 * 
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT.  Please see the
 * License for the specific language governing rights and limitations
 * under the License."
 * 
 * @APPLE_LICENSE_HEADER_END@
 */
/*

	user-level program to display and change debug variables in an 
	active AIX kernel running appletalk. Is is assumed that there
	are 2 u_long bit-mapped variables and that defines in DEBUG_H
	assign the bit values. Bits are set in one of 2 variables based
	on the define prefix of either D_M (for modules) or D_L
	(for types). The defines are displayed in an abbreviated format
	of L_xxx for D_L_xxx defines and M_xxx for D_M_xxx 
	defines.

	revision history
	----------------
	09-08-94	0.01 jjs Created
	09-12-94	0.02 jjs minor
	04-03-96	0.03 jjs allow running & updating of bits in file
						even if stack down

*/


#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <string.h>
#include <ctype.h>

#include <fcntl.h>
#include <at/appletalk.h>

#include <h/atlog.h>
#include <h/debug.h>
#include <h/lap.h>

#define FALSE	0
#define TRUE	1
			/* location of header file for 4.1 (and now that the
			   3.2 tree was put into the same structure, we don't
			   need separate defines i
			 */
#define DEBUG_H	"debug.h"	
#define BASE_PATH "/src/bos/kernext/atalk/h"
#define DELIMS		"	, "
#define COLUMNS		2
#define	MOD			0		/* array indices */
#define LEV			1
#define VERSION "setdbg v0.04 9-25-96"
#define HELP \
"	enter list of numbers to toggle the state of\n\
	an asterisk (*) marks bits which are active\n\
	numbers can be separated by spaces,tabs or commas\n\
	ranges of numbers are allowed (e.g. 5,9-15,20)\n\
	path to debug.h is required and is searched for in\n\
	the following order:\n\
		1. BASE env variable is set and valid\n\
		2. AT_DEBUG_DIR points to the dir where it is\n\
		3. debug.h is in the current directory\n\
	other commands:\n\
	q - quit without updating kernel\n\
	x - eXit and update kernel\n\
	t - toggle state of all bits\n\
	u - update kernel immediately\n\
	l - list status\n\
	s - set all bits to ON  c - clear all bits\n"



extern dbgBits_t dbgBits;

main(argc, argv)
	int argc;
	char **argv;
{

	char buf[80], instr[80];
	char defStr[40], *t, *s;
	u_long	defBytes[64];
	u_char	type[64];
	int i = 0,j,c,set;
	int found = 0;		/* set to TRUE if any lines found */
	int redisplay;
	int quit = 0;
	int lastRange, range=FALSE;

	FILE *fin;

	printf("%s\n",VERSION);
	if (kernelDbg(DBG_GET, &dbgBits)) {
		printf("error retreiving kernel dbg bit status\n");
	/*	exit(1); */ ;
	}
	if (argc > 1) {
		if (!(fin = fopen(argv[1],"r"))) {
			printf("error opening input file %s, aborting\n",argv[1]);
			exit(1);
		}
	} else {
		if ((t = getenv("BASE")) == NULL) {
	    	printf("env variable BASE not found, trying variable AT_DEBUG_DIR\n");
	    } else {
	     	sprintf(buf,"%s%s/%s",t,BASE_PATH,DEBUG_H);
	        t = buf;
			goto open;
	       	}

		if (t = getenv("AT_DEBUG_DIR")) {
	     	sprintf(buf,"%s/%s",t,DEBUG_H);
	        t = buf;
			goto open;
		} else {
	      	printf("env variable AT_DEBUG_DIR not found, trying current dir\n");
		}
		t = DEBUG_H;
open:
		if (access(t,R_OK)) {
			printf("%s not accessable, aborting\n",t);
			exit(1);
		}
		if (!(fin = fopen(t,"r"))) {
			printf("error opening input file %s, aborting\n",t);
			exit(1);
		}
	}
	do {								/* user input loop */
		i=0;
		while (!feof(fin)) {			/* loop to read & display file */
			if (!fgets(buf, sizeof(buf)-1, fin))
				break;
			if (sscanf(buf," #define D_L_%s 0x%lx", 
				defStr, &defBytes[i]) == 2){
				type[i] = LEV;
			}else {
				if (sscanf(buf," #define D_M_%s 0x%lx", 
					defStr, &defBytes[i]) == 2){
					type[i] = MOD;
				} else
					continue;
			}
			found = 1;
			set = type[i] == MOD ? dbgBits.dbgMod & defBytes[i] : dbgBits.dbgLev & defBytes[i];
			printf("%2d %c%c_%-20s",i+1, set ? '*' : ' ', type[i] == MOD ? 'M' : 'L', defStr);
			i++;
			if (!(i%COLUMNS))
				printf("\n");
		}
		if (!found) {
			printf("no defines found in %s. Aborting\n",DEBUG_H);
			exit(1);
		}
		if ((i%COLUMNS))
			printf("\n");
		rewind(fin);
		do {
			redisplay = 1;
			printf("cmd>");
			gets(instr);
			if (strlen(instr) == 0) {
				printf("enter ? for help\n");
				continue;
			}
			s = strtok(instr, DELIMS);
			switch ( s[0] = toupper(s[0])) {
				case 'X':					/* eXit and update kernel */
				case 'Q':					/* quit, don't update kernel */
					quit = 1;
					break;
				case 'S':					/* set all bits */
					dbgBits.dbgMod = 0xffffffff;
					dbgBits.dbgLev = 0xffffffff;
					break;
				case 'C':					/* clear all bits */
					dbgBits.dbgMod = 0;
					dbgBits.dbgLev = 0;
					break;
				case 'T':					/* toggle all bits */
					dbgBits.dbgMod ^= 0xffffffff;
					dbgBits.dbgLev ^= 0xffffffff;
					break;
				case 'U':					/* update kernel imediately */
					redisplay = 0;
					kernelDbg(DBG_SET, &dbgBits);
					break;
				case 'L':					/* list status */
					redisplay = 1;
					break;
				case '?':					/* help */
					redisplay = 0;
					printf(HELP);
					break;
				default:
					if (isdigit(s[0]))
						break;
					printf("unknown command, enter '?' for help\n");
					redisplay = 0;

					break;
			}
			if (quit || isdigit(s[0]) ||redisplay)
				break;
		} while (1);
		if (quit)
			break;
		if (!isdigit(s[0]))
			continue;
		do {
			c = atoi(s);
			if (c>=1 && c < i) {
				c--;
				if (t=strchr(s,'-')) {
					lastRange=c-1;
					t++;
					c = atoi(t);
					c--;
					range=TRUE;
				}
				if (range && lastRange <= c) {
					for(j=lastRange+1; j<= c; j++)
						if (type[j] == MOD)
							dbgBits.dbgMod ^= defBytes[j];
						else
							dbgBits.dbgLev ^= defBytes[j];
					range=FALSE;
				}
				else {
					lastRange = c;
					if (type[c] == MOD)
						dbgBits.dbgMod ^= defBytes[c];
					else
						dbgBits.dbgLev ^= defBytes[c];
				}
			}
			else if (c == 0)
				if (s[0] == '-') 
					range=TRUE;
		} while (s = strtok(NULL, DELIMS));

	} while (1);
	if (s[0] == 'X') 
		if (kernelDbg(DBG_SET, &dbgBits)) 
			printf("error updating kernel\n");
		else
			printf("kernel updated\n");
	else 
		printf(" leaving kernel unchanged\n");


	fclose(fin);
	exit(0);
}
